File: Java/lib/src/test/java/leetcode/AddTwoNumbers/AddTwoNumbersTest.java
package leetcode.AddTwoNumbers;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import java.util.stream.Stream;
import leetcode.Helpers.ListNode;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class AddTwoNumbersTest {
  protected Solution solution = new Solution();

  static Stream<Arguments> testsProvider() {
    return Stream.of(
        Arguments.of(
            "same number of digits",
            ListNode.fromArray(new int[] {2, 4, 3}),
            ListNode.fromArray(new int[] {5, 6, 4}),
            new int[] {7, 0, 8}),
        Arguments.of(
            "different number of digits",
            ListNode.fromArray(new int[] {2, 4, 3}),
            ListNode.fromArray(new int[] {5, 6}),
            new int[] {7, 0, 4}),
        Arguments.of(
            "one list is empty",
            ListNode.fromArray(new int[] {}),
            ListNode.fromArray(new int[] {5, 6}),
            new int[] {5, 6}),
        Arguments.of(
            "both lists are empty",
            ListNode.fromArray(new int[] {}),
            ListNode.fromArray(new int[] {}),
            new int[] {}));
  }

  @ParameterizedTest(name = "{0}")
  @MethodSource("testsProvider")
  void test(String name, ListNode l1, ListNode l2, int[] expected) {
    ListNode result = solution.addTwoNumbers(l1, l2);
    assertArrayEquals(expected, result != null ? result.toArray() : new int[] {});
  }
}

File: Go/add_two_numbers/add_two_numbers_test.go
package add_two_numbers

import (
	"leetcode/lib"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestAddTwoNumbers(t *testing.T) {
	cases := []struct {
		name     string
		l1       *ListNode
		l2       *ListNode
		expected []int
	}{
		{name: "same number of digits", l1: lib.NewList([]int{2, 4, 3}), l2: lib.NewList([]int{5, 6, 4}), expected: []int{7, 0, 8}},
		{name: "different number of digits", l1: lib.NewList([]int{2, 4, 3}), l2: lib.NewList([]int{5, 6}), expected: []int{7, 0, 4}},
		{name: "one list is empty", l1: lib.NewList([]int{}), l2: lib.NewList([]int{5, 6}), expected: []int{5, 6}},
		{name: "both lists are empty", l1: lib.NewList([]int{}), l2: lib.NewList([]int{}), expected: []int{}},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			actual := addTwoNumbers(c.l1, c.l2)
			assert.Equal(t, c.expected, actual.ToSlice())
		})
	}
}
